File: test/cpu/tst.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"
#include "test.hpp"

#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;
struct F_TST {
    F_TST() {
        // TST.B %D2
        TEST::SET_W(0, 0045000 | 2);
        TEST::SET_W(2, TEST_BREAK);

        // TST.W %D2
        TEST::SET_W(4, 0045100 | 2);
        TEST::SET_W(6, TEST_BREAK);

        // TST.L %D2
        TEST::SET_W(8, 0045200 | 2);
        TEST::SET_W(10, TEST_BREAK);
        jit_compile(0, 12);
    }};
BOOST_FIXTURE_TEST_SUITE(TST, Prepare, *boost::unit_test::fixture<F_TST>())


BOOST_AUTO_TEST_SUITE(Byte)

BOOST_AUTO_TEST_CASE(normal) {
    cpu.D[2] = 1;
    run_test(0);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.N);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 0x80;
    run_test(0);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0;
    run_test(0);
    BOOST_TEST(cpu.Z);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Word)

BOOST_AUTO_TEST_CASE(normal) {
    cpu.D[2] = 1;
    run_test(4);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.N);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 0x8000;
    run_test(4);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0;
    run_test(4);
    BOOST_TEST(cpu.Z);
}
BOOST_AUTO_TEST_SUITE_END()

BOOST_AUTO_TEST_SUITE(Long)

BOOST_AUTO_TEST_CASE(normal) {
    cpu.D[2] = 1;
    run_test(8);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.N);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 0x80000000;
    run_test(8);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0;
    run_test(8);
    BOOST_TEST(cpu.Z);
}
BOOST_AUTO_TEST_SUITE_END()

BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/tst.lua
-- TST.B %D2
MEM[0].W = OCT"045000" | 2
MEM[2].W = TEST.BREAK

-- TST.W %D3
MEM[4].W = OCT"045103" | 3
MEM[6].W = TEST.BREAK

-- TST.L %D4
MEM[8].W = OCT"045204" | 4
MEM[10].W = TEST.BREAK

JIT_COMPILE(0, 12)

local run = {}

run.B = {
    n = farray(TF, function(b)
            CPU.D[3].B = cond(b, 0x80, 0)
            TEST.RUN(0)
            TEST.CHECK ("CPU.CC.N", b)
            TEST.CHECK ("CPU.CC.V", false)
            TEST.CHECK ("CPU.CC.C", false)
        end),
    z = farray(TF, function(b)
            CPU.D[3].B = cond(b, 0, 1)
            TEST.RUN(0)
            TEST.CHECK ("CPU.CC.Z", b)
        end),
}

run.W = {
    n = farray(TF, function(b)
            CPU.D[4].W = cond(b, 0x8000, 0)
            TEST.RUN(4)
            TEST.CHECK ("CPU.CC.N", b)
            TEST.CHECK ("CPU.CC.V", false)
            TEST.CHECK ("CPU.CC.C", false)
        end),
    z = farray(TF, function(b)
            CPU.D[4].W = cond(b, 0, 1)
            TEST.RUN(4)
            TEST.CHECK ("CPU.CC.Z", b)
        end),
}

run.L = {
    n = farray(TF, function(b)
            CPU.D[5].L = cond(b, 0x80000000, 0)
            TEST.RUN(8)
            TEST.CHECK ("CPU.CC.N", b)
            TEST.CHECK ("CPU.CC.V", false)
            TEST.CHECK ("CPU.CC.C", false)
        end),
    z = farray(TF, function(b)
            CPU.D[5].L = cond(b, 0, 1)
            TEST.RUN(8)
            TEST.CHECK ("CPU.CC.Z", b)
        end),
}
return run